src/audio/selector/selector.c

```diff
@@ -743,3 +571,327 @@ UT_STATIC void sys_comp_selector_init(void)
 }
 
 DECLARE_MODULE(sys_comp_selector_init);
+#else
+static void build_config(struct comp_data *cd)
+{
+	enum sof_ipc_frame valid_format;
+
+	cd->source_format = cd->md.base_cfg.audio_fmt.depth;
+	audio_stream_fmt_conversion(cd->md.base_cfg.audio_fmt.depth,
+				    cd->md.base_cfg.audio_fmt.valid_bit_depth,
+				    &cd->source_format,
+				    &valid_format,
+				    cd->md.base_cfg.audio_fmt.s_type);
+
+	audio_stream_fmt_conversion(cd->md.output_format.depth,
+				    cd->md.output_format.valid_bit_depth,
+				    &cd->sink_format,
+				    &valid_format,
+				    cd->md.output_format.s_type);
+
+	cd->config.in_channels_count = cd->md.base_cfg.audio_fmt.channels_count;
+	cd->config.out_channels_count = cd->md.output_format.channels_count;
+}
+
+static int selector_init(struct processing_module *mod)
+{
+	struct module_data *md = &mod->priv;
+	struct module_config *cfg = &md->cfg;
+	struct comp_dev *dev = mod->dev;
+	struct comp_data *cd;
+	int ret;
+
+	comp_dbg(dev, "selector_init()");
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd)
+		return -ENOMEM;
+
+	md->private = cd;
+
+	ret = memcpy_s(&cd->md, sizeof(cd->md), cfg->data, sizeof(cd->md));
+	assert(!ret);
+
+	build_config(cd);
+
+	return 0;
+}
+
+static void set_selector_params(struct comp_dev *dev,
+				struct sof_ipc_stream_params *params)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer __sparse_cache *source;
+	struct comp_buffer __sparse_cache *sink;
+	struct comp_buffer *src_buf;
+	struct list_item *sink_list;
+	int i;
+
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK)
+		params->channels = cd->config.in_channels_count;
+	else
+		params->channels = cd->config.out_channels_count;
+
+	params->rate = cd->md.base_cfg.audio_fmt.sampling_frequency;
+	params->frame_fmt = cd->source_format;
+
+	for (i = 0; i < SOF_IPC_MAX_CHANNELS; i++)
+		params->chmap[i] = (cd->md.output_format.ch_map >> i * 4) & 0xf;
+
+	/* update each sink format */
+	list_for_item(sink_list, &dev->bsink_list) {
+		struct comp_buffer *sink_buf;
+		struct ipc4_audio_format *out_fmt;
```

Comment:
same comment - I'd move the `sink` declaration here too, and possibly initialised variables here.